### PR TITLE
fix: seclang scanner mis-parsing escaped quotes right after a macro

### PR DESCRIPTION
**Pull request by fzipi**
## Summary

An action value with an escaped quote immediately after a `%{VARIABLE}` macro expansion fails to parse, e.g.:

```
SecAction "id:1,phase:5,log,pass,logdata:'\'Field\' : \'%{TX.foo}\''"
```

fails with:

```
Rules error. ... Expecting an action, got: 'Field' : '%{TX.foo}''"
```

This was found via a real plugin rule in [coreruleset/traffic-observation-plugin](https://github.com/coreruleset/traffic-observation-plugin), which fails to load under nginx (`owasp/modsecurity-crs` Docker image, which links `libmodsecurity` v3):

https://github.com/coreruleset/traffic-observation-plugin/actions/runs/35602944525/job/106343176175?pr=7

```
nginx: [emerg] "modsecurity_rules_file" directive Rules error. File: .../traffic-observation-after.conf.
Line: 107. Column: 877. Expecting an action, got:  'Method' : '%{REQUEST_METHOD}', 'LenFilename' : ...
```

### Root cause

In `src/parser/seclang-scanner.ll`, `FREE_TEXT_QUOTE_MACRO_EXPANSION` (and its double-quote counterpart) recognize `\'` as a literal escaped quote only through alternatives that bundle in a preceding non-backslash "anchor" character as part of the same token match, e.g. `[^\\][\\][']`.

When a `%{VARIABLE}` macro closes, the closing `}` is consumed by its own rule in `EXPECTING_ACTION_PREDICATE_VARIABLE`, which switches state without emitting a token for it. That `}` is therefore not available as the anchor for the very next token. If that next token starts with `\'` (an escaped quote right after the macro), none of the anchor-based alternatives can match it — the scanner falls back to matching just the lone `\` as ordinary text, then treats the following bare `'` as the real closing quote, terminating the value early and leaving the rest of the action list unparseable. The same gap exists at the very start of a quoted value (no anchor exists before the first character either).

### Fix

Add two new "leading escape" macros that match an escaped quote (`\'` / `\"`) at the very start of a token, optionally followed by more ordinary free text:

```
FREE_TEXT_QUOTE_MACRO_EXPANSION_LEADING_ESCAPE          [\\][']{FREE_TEXT_QUOTE_MACRO_EXPANSION}?
FREE_TEXT_DOUBLE_QUOTE_MACRO_EXPANSION_LEADING_ESCAPE   [\\]["]{FREE_TEXT_DOUBLE_QUOTE_MACRO_EXPANSION}?
```

and wire them into every scanner state that consumes the corresponding free-text macro:
- `ACTION_PREDICATE_ENDS_WITH_QUOTE` / `ACTION_PREDICATE_ENDS_WITH_DOUBLE_QUOTE` (action values, e.g. `msg`, `logdata`, `tag`, `ver`)
- `NO_OP_INFORMED_ENDS_WITH_QUOTE` / `EXPECTING_PARAMETER_ENDS_WITH_QUOTE` (double-quoted operator parameters)

`src/parser/seclang-scanner.cc` was regenerated from the updated `.ll` via `flex` (the large diff there is normal DFA-table churn from regeneration, not hand-edited).

Scope note: the same anchor-character issue likely also exists in the `setvar` value tokenizer (`FREE_TEXT_EQUALS_QUOTE_MACRO_EXPANSION` / `FREE_TEXT_EQUALS_MACRO_EXPANSION`), but that's a structurally different state machine and wasn't part of the reported/reproduced failure, so it's left out of this PR to keep the change scoped and reviewable.

## Test plan

- Added `test/test-cases/regression/misc-escaped-quote-after-macro-expansion.json` with two cases (single-quote and double-quote value delimiters) that reproduce the exact failure pattern and assert the rule loads and the macro-substituted `logdata` value appears correctly (and in full) in the error log.
- Verified both new tests **fail** against the pre-fix scanner with the exact reported parser error, and **pass** with the fix.
- Ran the full regression suite: `720 passed, 0 failed, 13 skipped` (skips are pre-existing, feature-gated).
- `test/unit_tests` has 11 pre-existing failures unrelated to this change (`geoLookup`/`rbl` network-dependent tests and a `phpArgsNames` transformation edge case in the `secrules-language-tests` submodule), unaffected by this fix.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed `logdata` action values being truncated when escaped single or double quotes follow a macro expansion.
  * Correctly preserves quotes preceded by odd-length backslash sequences after macro expansion.
  * Preserved complete escaped error-log messages in affected request-processing scenarios.

* **Tests**
  * Added regression coverage for escaped quotes following request method and protocol macro expansions.
  * Added coverage for single and double quotes preceded by three backslashes.

### CODE REVIEW

File: src/parser/seclang-scanner.ll [hunk 1]
```diff
@@ -442,6 +442,15 @@ DOUBLE_QUOTE_BUT_SCAPED                 (")
...
 FREE_TEXT_DOUBLE_QUOTE_MACRO_EXPANSION          ((([^"%])|([%][^{]))|([^\\][\\][%][{])|([^\\]([\\][\\])+[\\][%][{])|[^\\][\\]["]|[^\\]([\\][\\])+[\\]["])+
+
+/* An escaped quote at the very start of a fresh token has no preceding
+ * character to serve as the "not a backslash" anchor the alternatives
+ * above rely on. That happens right after a %{VARIABLE} macro closes
+ * (the closing '}' is consumed by its own rule, in EXPECTING_ACTION_PREDICATE_VARIABLE,
+ * and is not available to the next token) and at the very start of a
+ * quoted value. These two macros cover that leading-escape case. */
+FREE_TEXT_QUOTE_MACRO_EXPANSION_LEADING_ESCAPE          [\\][']{FREE_TEXT_QUOTE_MACRO_EXPANSION}?
+FREE_TEXT_DOUBLE_QUOTE_MACRO_EXPANSION_LEADING_ESCAPE   [\\]["]{FREE_TEXT_DOUBLE_QUOTE_MACRO_EXPANSION}?
```
[the remark this answers is not in the document]
Comment (fzipi) [reply src/parser/seclang-scanner.ll]: Good catch — applied your suggested fix (`[\\]([\\][\\])*[']`/`[\\]([\\][\\])*["]`) verbatim in de093f34, plus two new regression cases (single- and double-quote, 3-backslash run right after a macro) that I verified fail on the old single-backslash-only pattern and pass with the fix. Full regression_tests suite still green (722 passed, 0 failed).

File: src/parser/seclang-scanner.ll [hunk 2]
```diff
@@ -442,6 +442,17 @@ DOUBLE_QUOTE_BUT_SCAPED                 (")
...
+/* An escaped quote at the very start of a fresh token has no preceding
+ * character to serve as the "not a backslash" anchor the alternatives
+ * above rely on. That happens right after a %{VARIABLE} macro closes
+ * (the closing '}' is consumed by its own rule, in EXPECTING_ACTION_PREDICATE_VARIABLE,
+ * and is not available to the next token) and at the very start of a
+ * quoted value. These two macros cover that leading-escape case: an odd
+ * run of backslashes (1, 3, 5, ...) before the quote still escapes it,
+ * matching the pairing convention used by the alternatives above. */
+FREE_TEXT_QUOTE_MACRO_EXPANSION_LEADING_ESCAPE          [\\]([\\][\\])*[']{FREE_TEXT_QUOTE_MACRO_EXPANSION}?
+FREE_TEXT_DOUBLE_QUOTE_MACRO_EXPANSION_LEADING_ESCAPE   [\\]([\\][\\])*["]{FREE_TEXT_DOUBLE_QUOTE_MACRO_EXPANSION}?
```
[the remark this answers is not in the document]
Comment (fzipi) [reply src/parser/seclang-scanner.ll:457]: Checked this empirically before responding: the committed `src/parser/seclang-scanner.cc` in this PR *is* the regenerated file with the leading-escape rules — I ran `flex` on `.ll` and committed the result each time, same as the file's existing convention (it's listed directly in `libmodsecurity_la_SOURCES` at `src/Makefile.am`, compiled with plain g++/gcc, no flex step in the default path per `configure.ac:286`).

To be sure, I did a clean-room check: checked out this exact PR commit, touched only `seclang-scanner.cc` (no `flex` invocation at all) to force a from-scratch `g++` recompile via `make -C src`, then ran the regression tests against that build. All pass, including the odd-backslash-run cases. So a normal user building without `--enable-parser-generation` does get the fix from this diff as-is — no additional file is missing.

If you're seeing a different result, could you share the exact command? Happy to dig further, but I can't reproduce a build where the fix is absent.

[the remark this answers is not in the document]
Comment (fzipi) [reply src/parser/seclang-scanner.ll:457]: Good catch, this was a real gap — confirmed with the exact repro you gave (`msg:prefix%{REQUEST_METHOD}\"suffix`): it failed with `Rules error. ... Invalid input:  suffix"`. Fixed in 237bc6e2 by adding `FREE_TEXT_COMMA_DOUBLE_QUOTE_MACRO_EXPANSION_LEADING_ESCAPE` and wiring it into `ACTION_PREDICATE_ENDS_WITH_COMMA_OR_DOUBLE_QUOTE`, same pattern as the other two states. Added a regression case, verified it fails on the previous two-state fix and passes with this one, full regression_tests suite green (723 passed, 0 failed).

**Comment by airween** [comment]
Hi @fzipi,

thank you for this PR.

Notes/questions:
* could you check which version of `flex` you used? The generated [code](https://github.com/fzipi/ModSecurity/blob/237bc6e2f887b5b8453832a8a3b1a399d8de6d11/src/parser/seclang-scanner.cc#L18-L20) shows `2.6.4`, which is I think the latest, so that's the expected (just to be sure)
* your [new test](https://github.com/fzipi/ModSecurity/blob/237bc6e2f887b5b8453832a8a3b1a399d8de6d11/test/test-cases/regression/misc-escaped-quote-after-macro-expansion.json) was not checked in CI (with `make check`); the reason is simple: all test must be exists in `test-suite.in`, this one somewhere around [here](https://github.com/fzipi/ModSecurity/blob/237bc6e2f887b5b8453832a8a3b1a399d8de6d11/test/test-suite.in#L82); (you can run `cd test; ./test-suite.sh` if you are in a clean directory, without your own tests)

Thank you again!

**Comment by fzipi** [comment]
Hi @airween, thanks for the review!

- flex version: confirmed, `flex --version` gives `2.6.4` locally, matching the generated header — so that's expected.
- Test registration: you were right, the new test json existed but was never added to `test-suite.in`, so it never ran in CI. Fixed in 91c29339 — added `TESTS+=test/test-cases/regression/misc-escaped-quote-after-macro-expansion.json`. Verified locally with `make check` (all 5 assertions in the new test pass; ran the full suite too, remaining failures are pre-existing and due to optional deps — Lua/GeoIP/LibCURL/LibXML2 — not enabled in my local build, unrelated to this change).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Comment by airween** [comment]
> * flex version: confirmed, `flex --version` gives `2.6.4` locally, matching the generated header — so that's expected.

thanks - I just asked because I see [this](https://github.com/owasp-modsecurity/ModSecurity/pull/3641/changes#diff-6d6ec826decbc1120d98a0c917b7440fe53537e225d13532711550ae13c4627bR199) line:

```C++
extern yy_size_t yyleng;
``` 
I re-generated this in local, and got this:

```C++
extern int yyleng;
```
which is the same as in the diff above, on the left side.

It would be nice to figure out, why is this diff there.

Note, I got the same `flex` version:
```
$ flex --version
flex 2.6.4
```
> * Test registration: you were right, the new test json existed but was never added to `test-suite.in`, so it never ran in CI. Fixed in [91c2933](https://github.com/owasp-modsecurity/ModSecurity/commit/91c293397ac3a0612f8580e4dbdf0b7faaa63918) — added `TESTS+=test/test-cases/regression/misc-escaped-quote-after-macro-expansion.json`. Verified locally with `make check` (all 5 assertions in the new test pass; ran the full suite too, remaining failures are pre-existing and due to optional deps — Lua/GeoIP/LibCURL/LibXML2 — not enabled in my local build, unrelated to this change).

I see that both in [CI](https://github.com/owasp-modsecurity/ModSecurity/actions/runs/35864688181/job/107193180432?pr=3641#step:11:158), and in the [diff](https://github.com/owasp-modsecurity/ModSecurity/pull/3641/changes#diff-09fd64e3341e336cb5c6641c75a8e5c4541f8990a2db3833e18f67f8c3daf2f8R81). Excellent, thank you!
